src/python/pants/engine/internals/engine_test.py

```diff
@@ -561,7 +568,61 @@ def a_rule(n: int) -> Output:
             item for item in finished if item["name"] == "pants.engine.internals.engine_test.a_rule"
         )
         artifacts = workunit["artifacts"]
-        assert artifacts["some_arbitrary_key"] == EMPTY_DIGEST
+        assert artifacts["some_arbitrary_key"] == EMPTY_SNAPSHOT
+
+
+@dataclass(frozen=True)
+class Output(EngineAwareReturnType):
+    val: Snapshot
+
+    def artifacts(self):
+        return {"some_arbitrary_key": self.val}
+
+
+@rule(desc="a_rule")
+def a_rule(snapshot: Snapshot) -> Output:
+    return Output(val=snapshot)
+
+
+class MoreComplicatedEngineAware(TestBase, SchedulerTestBase):
```

Comment:
I don't think `SchedulerTestBase` is being used. Can you see if it still passes when deleting?

If it can be deleted, bonus that we're trying to stop using `TestBase` to use `Pytest` style instead through the new `RuleRunner`. If you're up for it, it'd be great to rewrite.

First, make this a top-level function rather than a method. At the top of the function, add this:

```python
from pants.testutil.rule_runner import RuleRunner

...

def test_foo() -> None:
   rule_runner = RuleRunner(rules=[a_rule, QueryRule(...), ...])
```

Then, instead of `self.request_product`, use `rule_runner.request_product`.

Because the `RuleRunner` is created inside the test, you can then move all your dataclass and rule definitions to be inside the test function, at the top, above your `RuleRunner` setup. This results in better namespacing.

Reply:
This second bit is a little complicated to do since I'm using `self.scheduler` here; I'd rather save this cleanup for another PR

Reply:
That'd be `rule_runner.scheduler` instead of `self.scheduler`. It's fine to not do the second part, but preferred to fix because `TestBase` is deprecated.